### ISSUE TITLE
Unseen span labels in constituency parser model

### ISSUE THREAD

**Issue by zplizzi** [allenai/allennlp#1752]
I'm evaluating the constituency parser model on the PTB fragment distributed with NLTK (which I compiled into one file with all the merged trees).

However, I believe I ran into an issue marked in a "TODO" in the code [here](https://github.com/allenai/allennlp/blob/4920249af097d1557be6ece3e5ce80d1f6ea0333/allennlp/data/dataset_readers/penn_tree_bank.py#L184):

```
TODO(Mark): If we encounter a gold nested labelling at test time
which we haven't encountered, we won't be able to run the model
at all.
```

Specifically, the issue is when there is a span label in the evaluation data that wasn't in the training data. When this occurs, the model throws an error because OOV labels are dis-allowed. In my case the span label was "SQ-FRAG", indicating nested "SQ" and "FRAG" spans. I'm guessing this has to do with using the NLTK version of PTB instead of the LDC version (which isn't freely available).

I fixed the error by changing the span label namespace to something without the suffix "label" (ie "test"), which I understand enables OOV tokens to be legal. The model evaluates and returns the following results:

```
2018-09-11 22:47:41,597 - INFO - allennlp.commands.evaluate - Finished evaluating.
2018-09-11 22:47:41,597 - INFO - allennlp.commands.evaluate - Metrics:
2018-09-11 22:47:41,597 - INFO - allennlp.commands.evaluate - tag_accuracy: 0.01969830953647653
2018-09-11 22:47:41,597 - INFO - allennlp.commands.evaluate - evalb_recall: 0.9104670632597776
2018-09-11 22:47:41,597 - INFO - allennlp.commands.evaluate - evalb_precision: 0.9092911330451097
2018-09-11 22:47:41,597 - INFO - allennlp.commands.evaluate - evalb_f1_measure: 0.9098787182085938
```

I'm not exactly sure what it should output, but I suspect that "tag_accuracy" shouldn't be that low, so I suspect my attempted "fix" broke something else. Or perhaps the NLTK dataset is sufficiently different that this won't work at all. I would appreciate if someone knew if there was an easy fix to this issue or not.

**Comment by DeNeutoy** [comment]
I'll leave this here:
https://github.com/mitchellstern/minimal-span-parser

Caveat - I don't know that what you might find there is _identical_ to the data we used to train on - I think possibly the pos-tags might be predicted there, rather than gold. But it's certainly closer.

Also your method for tweaking the format won't work, because it means that all the tags will be UNK, as the dataset reader assumes the vocabulary of the labels has a particular name. 

It's a bit tricky to allow unseen labels at test time - it causes all sorts of problems, because it means that it's really easy to silently evaluate on a dataset which has unseen labels, which is probably even worse than crashing. It is particularly annoying in the case of syntactic parsing, because there are about a billion different formats and tag labels, so it seems like in some cases this is desirable.

Hopefully that's helpful!

**Comment by zplizzi** [comment]
Much appreciated.

For anyone else trying this data, I had to change the string "VROOT" to "TOP" in [this line](https://github.com/allenai/allennlp/blob/4920249af097d1557be6ece3e5ce80d1f6ea0333/allennlp/data/dataset_readers/penn_tree_bank.py#L72). This is the top-level span label which appears to be different in this data vs the LDC one.

And here's the evaluation results on the test set (23):
```
2018-09-12 00:46:46,773 - INFO - allennlp.commands.evaluate - Finished evaluating.
2018-09-12 00:46:46,774 - INFO - allennlp.commands.evaluate - Metrics:
2018-09-12 00:46:46,774 - INFO - allennlp.commands.evaluate - tag_accuracy: 0.9938221790664811
2018-09-12 00:46:46,774 - INFO - allennlp.commands.evaluate - evalb_recall: 0.931543048152498
2018-09-12 00:46:46,774 - INFO - allennlp.commands.evaluate - evalb_precision: 0.9432387312186978
2018-09-12 00:46:46,774 - INFO - allennlp.commands.evaluate - evalb_f1_measure: 0.9373544083724419
```

Ever so slightly lower score than quoted for this model but plenty good enough to work with. Thanks!
